message_verifier.go

```diff
@@ -0,0 +1,272 @@
+	return nil
+}
+
+// validateRequiredHeaders ensures all required headers exist in the signature headers.
+func (ma *messageAttributes) validateRequiredHeaders(requiredHeaders []string) error {
+	// if there are no required headers, don't do anything.
+	if len(requiredHeaders) == 0 {
+		return nil
+	}
+
+	headers := map[string]struct{}{}
+
+	for _, header := range ma.signatureHeaders {
+		headers[header] = struct{}{}
+	}
+
+	missingHeaders := []string{}
+
+	for _, header := range requiredHeaders {
+		if _, ok := headers[header]; !ok {
+			missingHeaders = append(missingHeaders, header)
+		}
+	}
+
+	if len(missingHeaders) > 0 {
+		return NewSignatureError(fmt.Sprintf("signature missing required headers: %s", strings.Join(missingHeaders, ", ")), nil)
+	}
+
+	return nil
+}
+
+// verifySignature verifies that the signature is valid for the headers.
+func (ma *messageAttributes) verifySignature(verifier Verifier) error {
+	signatureSum, err := base64.StdEncoding.DecodeString(ma.signature)
+	if err != nil {
+		return NewDataError("failed base64 decoding signature", err)
+	}
+
+	signatureString, err := generateSignatureString(ma.req, ma.signatureHeaders)
+	if err != nil {
+		return err
+	}
+
+	return verifier.Verify(signatureSum, signatureString)
+}
+
+// getHashingAlgorithmByDigestKey finds the hashing algorithm for the given key in a digest header value.
+func getHashingAlgorithmByDigestKey(digestKey string) (*HashingAlgorithm, error) {
+	switch digestKey {
+	case crypto.SHA256.String():
+		return fetchCryptoHash(crypto.SHA256)
+	case crypto.SHA512.String():
+		return fetchCryptoHash(crypto.SHA512)
+	}
+
+	return nil, NewDataError(fmt.Sprintf("digest algorithm could not be found or is not supported: %s", digestKey), nil)
+}
+
+// determineSignature determines whether the signature belongs in the Signature or the
+// Authorization header. If both or neither contain a signature then an error is returned.
+// Signatures that contains a duplicate field, e.g. 2 fields named 'keyId', should not be
+// evaluated as valid.
+func determineSignature(headers http.Header) (string, error) {
+	validSignatureHeader := false
+	validAuthorizationHeader := false
+
+	signatureHeader := headers.Get(string(Signature))
+	if signatureHeader != "" {
+		validSignatureHeader = strings.Count(signatureHeader, signatureHeaderKeyIDKey) == 1 &&
+			strings.Count(signatureHeader, signatureHeaderAlgorithmKey) == 1 &&
+			strings.Count(signatureHeader, signatureHeaderHeadersKey) == 1 &&
+			strings.Count(signatureHeader, signatureHeaderSignatureKey) == 1
+	}
+
+	authorizationHeader := headers.Get(string(Authorization))
+	if authorizationHeader != "" {
+		validAuthorizationHeader = strings.Count(authorizationHeader, signatureHeaderKeyIDKey) == 1 &&
+			strings.Count(authorizationHeader, signatureHeaderAlgorithmKey) == 1 &&
+			strings.Count(authorizationHeader, signatureHeaderHeadersKey) == 1 &&
+			strings.Count(authorizationHeader, signatureHeaderSignatureKey) == 1
+	}
+
+	switch {
```

Comment:
It may be nitpicking, but if someone makes a request that contains both signature and authorization headers, but one of them would be invalid (for example would contain 2 `signatureHeaderKeyIDKey`) we would still proceed with the valid one. I'm not sure if it's desired though

Reply:
Hmm, I think if one is valid and one is invalid, regardless of what makes it invalid, there is no ambiguity on which to use. I guess if both are valid, we could check to see if they are identical and proceed. The RFC doesn't specify anything regarding this so it's up to us on implementation

Reply:
In that case feel free to ignore me ;)